src/Evolution/DiscontinuousGalerkin/Actions/NormalVectorQuantities.hpp

```diff
@@ -0,0 +1,176 @@
+// Distributed under the MIT License.
+// See LICENSE.txt for details.
+
+#pragma once
+
+#include <cstddef>
+#include <memory>
+#include <optional>
+#include <unordered_map>
+
+#include "DataStructures/DataBox/DataBox.hpp"
+#include "DataStructures/DataBox/DataBoxTag.hpp"
+#include "DataStructures/DataVector.hpp"
+#include "DataStructures/Tensor/EagerMath/DotProduct.hpp"
+#include "DataStructures/Tensor/EagerMath/Magnitude.hpp"
+#include "DataStructures/Tensor/Tensor.hpp"
+#include "DataStructures/Variables.hpp"
+#include "Domain/CoordinateMaps/CoordinateMap.hpp"
+#include "Domain/CoordinateMaps/Tags.hpp"
+#include "Domain/Structure/Direction.hpp"
+#include "Evolution/DiscontinuousGalerkin/NormalVectorTags.hpp"
+#include "Utilities/Gsl.hpp"
+#include "Utilities/TMPL.hpp"
+#include "Utilities/TypeTraits/CreateHasTypeAlias.hpp"
+
+namespace evolution::dg::Actions::detail {
+CREATE_HAS_TYPE_ALIAS(inverse_spatial_metric_tag)
+CREATE_HAS_TYPE_ALIAS_V(inverse_spatial_metric_tag)
+
+template <bool HasInverseSpatialMetricTag = false>
+struct inverse_spatial_metric_tag_impl {
+  template <typename System>
+  using f = tmpl::list<>;
+};
+
+template <>
+struct inverse_spatial_metric_tag_impl<true> {
+  template <typename System>
+  using f = tmpl::list<typename System::inverse_spatial_metric_tag>;
+};
+
+template <typename System>
+using inverse_spatial_metric_tag = typename inverse_spatial_metric_tag_impl<
+    has_inverse_spatial_metric_tag_v<System>>::template f<System>;
+
+struct OneOverNormalVectorMagnitude : db::SimpleTag {
+  using type = Scalar<DataVector>;
+};
+
+template <size_t Dim>
+struct NormalVector : db::SimpleTag {
+  using type = tnsr::I<DataVector, Dim, Frame::Inertial>;
+};
+
+/*!
+ * \brief Computes the normal covector, magnitude of the unnormalized normal
+ * covector, and in curved spacetimes the normal vector.
+ */
+template <typename System, size_t Dim, typename FieldsOnFaceTags>
+void unit_normal_vector_and_covector(
+    const gsl::not_null<Scalar<DataVector>*> face_normal_magnitude,
+    const gsl::not_null<tnsr::i<DataVector, Dim, Frame::Inertial>*>
+        unit_normal_covector,
+    const gsl::not_null<Variables<FieldsOnFaceTags>*> fields_on_face,
+    const tnsr::i<DataVector, Dim, Frame::Inertial>&
+        unnormalized_normal_covector) noexcept {
+  if constexpr (has_inverse_spatial_metric_tag_v<System>) {
+    using inverse_spatial_metric_tag =
+        typename System::inverse_spatial_metric_tag;
+    auto& normal_vector = get<NormalVector<Dim>>(*fields_on_face);
+    const auto& inverse_spatial_metric =
+        get<inverse_spatial_metric_tag>(*fields_on_face);
+    // Compute unnormalized normal vector
+    for (size_t i = 0; i < Dim; ++i) {
```

Comment:
why not raise_or_lower_index()?

Reply:
because that's in GR and this doesn't really depend on GR